### PR TITLE
Demo for Callback-group-level Executor concept

### PR DESCRIPTION
**Pull request by ralph-lange**
Demo and test bench for use of the refined interface of the Executor (cf. https://github.com/ros2/rclcpp/pull/1218) as discussed in https://github.com/ros2/rclcpp/pull/1218#issuecomment-701181925.

### CODE REVIEW

File: cbg_executor_demo/CPPLINT.cfg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+set noparent
+linelength=100
+filter=-build/include_order
```
Comment (Karsten1987) [inline cbg_executor_demo/CPPLINT.cfg:3]: there's a newline missing.

File: cbg_executor_demo/include/cbg_executor_demo/ping_node.hpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+struct RTTData
+{
+  explicit RTTData(rclcpp::Time sent)
+  : sent_(sent) {}
+  rclcpp::Time sent_{0, 0};
+  rclcpp::Time high_received_{0, 0};
+  rclcpp::Time low_received_{0, 0};
+};
+
```
Comment (Karsten1987) [inline cbg_executor_demo/include/cbg_executor_demo/ping_node.hpp:38]:
```suggestion
```

File: cbg_executor_demo/include/cbg_executor_demo/ping_node.hpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+#define CBG_EXECUTOR_DEMO__PING_NODE_HPP_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <rclcpp/rclcpp.hpp>
+#include <std_msgs/msg/int32.hpp>
+
```
Comment (Karsten1987) [inline cbg_executor_demo/include/cbg_executor_demo/ping_node.hpp:25]:
```suggestion
```

File: cbg_executor_demo/include/cbg_executor_demo/ping_node.hpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace cbg_executor_demo
+{
+
+struct RTTData
+{
+  explicit RTTData(rclcpp::Time sent)
```
Comment (Karsten1987) [inline cbg_executor_demo/include/cbg_executor_demo/ping_node.hpp:32]: does pass by const ref make sense here?
```suggestion
  explicit RTTData(const rclcpp::Time & sent)
```

File: cbg_executor_demo/package.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  <maintainer email="ralph.lange@de.bosch.com">Ralph Lange</maintainer>
+  <license>Apache License 2.0</license>
+
+  <buildtool_depend>ament_cmake</buildtool_depend>
+
+  <build_depend>rclcpp</build_depend>
+  <build_depend>std_msgs</build_depend>
+
+  <exec_depend>rclcpp</exec_depend>
+  <exec_depend>std_msgs</exec_depend>
```
Comment (Karsten1987) [inline cbg_executor_demo/package.xml:16]:
```suggestion
  <depend>rclcpp</depend>
  <depend>std_msgs</depend>
```

File: cbg_executor_demo/src/cbg_executor_demo/parameter_helper.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) 2020 Robert Bosch GmbH
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/parameter_helper.cpp:1]: nitpick here, but do we really need a dedicated library for this one function? Why not just declare it `inline` in the header?

File: cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+#include "cbg_executor_demo/ping_node.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <functional>
+#include <memory>
+
+#include <cbg_executor_demo/parameter_helper.hpp>
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp:23]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  ping_timer_ = create_wall_timer(ping_period, std::bind(&PingNode::send_ping, this));
+  high_ping_publisher_ = create_publisher<Int32>("high_ping", rclcpp::SensorDataQoS());
+  low_ping_publisher_ = create_publisher<Int32>("low_ping", rclcpp::SensorDataQoS());
+
+  high_pong_subscription_ = create_subscription<Int32>(
+    "high_pong", rclcpp::SensorDataQoS(), std::bind(&PingNode::high_pong_received, this, _1));
+  low_pong_subscription_ = create_subscription<Int32>(
+    "low_pong", rclcpp::SensorDataQoS(), std::bind(&PingNode::low_pong_received, this, _1));
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp:46]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+void PingNode::send_ping()
+{
+  std_msgs::msg::Int32 msg;
+  msg.data = rtt_data_.size();
+  rtt_data_.push_back(RTTData(now()));
+  high_ping_publisher_->publish(msg);
+  low_ping_publisher_->publish(msg);
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp:56]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+void PingNode::high_pong_received(const std_msgs::msg::Int32::SharedPtr msg)
+{
+  rtt_data_[msg->data].high_received_ = now();
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp:62]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+void PingNode::low_pong_received(const std_msgs::msg::Int32::SharedPtr msg)
+{
+  rtt_data_[msg->data].low_received_ = now();
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/ping_node.cpp:68]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cbg_executor_demo/pong_node.hpp"
+
+#include <cassert>
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:18]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+{
+
+PongNode::PongNode()
+: rclcpp::Node("pong_node")
+{
+  using std::placeholders::_1;
+  using std_msgs::msg::Int32;
+
+  declare_parameter<double>("high_busyloop", 0.01);
+  high_pong_publisher_ = create_publisher<Int32>("high_pong", rclcpp::SensorDataQoS());
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:34]: other demos use the `this` pointer explicitly.
```suggestion
  high_pong_publisher_ = this->create_publisher<Int32>("high_pong", rclcpp::SensorDataQoS());
```

same below.

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+  declare_parameter<double>("low_busyloop", 0.01);
+  low_pong_publisher_ = create_publisher<Int32>("low_pong", rclcpp::SensorDataQoS());
+  rclcpp::SubscriptionOptionsWithAllocator<std::allocator<void>> options;
+  options.callback_group = second_cb_group;
+  low_ping_subscription_ = create_subscription<Int32>(
+    "low_ping", rclcpp::SensorDataQoS(),
+    std::bind(&PongNode::low_ping_received, this, _1), options);
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:50]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+rclcpp::CallbackGroup::SharedPtr PongNode::get_high_prio_callback_group()
+{
+  return get_callback_groups()[0].lock();  // ... the default callback group.
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:56]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+rclcpp::CallbackGroup::SharedPtr PongNode::get_low_prio_callback_group()
+{
+  return get_callback_groups()[1].lock();  // ... the second callback group created in the ctor.
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:62]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+
+void PongNode::high_ping_received(const std_msgs::msg::Int32::SharedPtr msg)
+{
+  std::chrono::nanoseconds busyloop = get_nanos_from_secs_parameter(this, "high_busyloop");
+  burn_cpu_cycles(busyloop);
+  high_pong_publisher_->publish(*msg);
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:70]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+void PongNode::low_ping_received(const std_msgs::msg::Int32::SharedPtr msg)
+{
+  std::chrono::nanoseconds busyloop = get_nanos_from_secs_parameter(this, "low_busyloop");
+  burn_cpu_cycles(busyloop);
+  low_pong_publisher_->publish(*msg);
+}
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:78]:
```suggestion
```

File: cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+      std::chrono::seconds{startTimeP.tv_sec} +
+    std::chrono::nanoseconds{startTimeP.tv_nsec};
+    int x = 0;
+    bool doAgain = true;
+    while (doAgain) {
+      while (x != std::rand() && x % 1000 != 0) {
+        x++;
+      }
+      timespec currentTimeP;
+      clock_gettime(clockId, &currentTimeP);
```
Comment (Karsten1987) [inline cbg_executor_demo/src/cbg_executor_demo/pong_node.cpp:97]: I suspect this won't work on OSX or MSVC, but let's see.

File: cbg_executor_demo/src/main.cpp [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+#include <windows.h>
+#endif
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/executor.hpp>
+
+#include "cbg_executor_demo/ping_node.hpp"
+#include "cbg_executor_demo/pong_node.hpp"
+
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/main.cpp:38]:
```suggestion
```

File: cbg_executor_demo/src/main.cpp [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+#include <vector>
+
+#ifndef _WIN32  // i.e., POSIX platform.
+#include <pthread.h>
+#else  // i.e., Windows platform.
+#include <windows.h>
+#endif
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/executor.hpp>
```
Comment (Karsten1987) [inline cbg_executor_demo/src/main.cpp:33]: alpha sort

```suggestion
#include <rclcpp/executor.hpp>
#include <rclcpp/rclcpp.hpp>
```

File: cbg_executor_demo/src/main.cpp [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+  if (cpu_id >= 0) {
+    DWORD_PTR cpuset = 1;
+    cpuset <<= cpu_id;
+    success &= (SetThreadAffinityMask(thread_handle, cpuset) != 0);
+  }
+#endif
+  return success;
+}
+
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/main.cpp:74]:
```suggestion
```

File: cbg_executor_demo/src/main.cpp [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+  ULARGE_INTEGER user_time;
+  user_time.LowPart = user_filetime.dwLowDateTime;
+  user_time.HighPart = user_filetime.dwHighDateTime;
+  std::chrono::nanoseconds t(100);  // Unit in FILETIME is 100ns.
+  t *= (kernel_time.QuadPart + user_time.QuadPart);
+  return t;
+#endif
+}
+
+
```
Comment (Karsten1987) [inline cbg_executor_demo/src/main.cpp:104]:
```suggestion
```

**Review by Karsten1987 — COMMENTED** [review]
mainly style nitpicks here.

One question I'd have tho is why we need a dedicated package for this. Would it make sense to integrate it in the `demo_nodes_cpp` package with a subfolder `executors`? I could see that there's more than one example for different executors to come in the future.

**Comment by clalancette** [comment]
> One question I'd have tho is why we need a dedicated package for this. Would it make sense to integrate it in the `demo_nodes_cpp` package with a subfolder `executors`? I could see that there's more than one example for different executors to come in the future.

Actually, we already have some executor examples over at https://github.com/ros2/examples/tree/master/rclcpp/executors .  So we may consider moving this over there instead.

**Comment by ralph-lange** [comment]
I'll work on @Karsten1987's findings first and then move the package as proposed by @clalancette to not disturb GitHub's resolve mechanism.

**Comment by ralph-lange** [comment]
Worked on all comments. Will open new PR for ros2/examples tomorrow.

**Pull request closed without merging**